cpp/src/arrow/util/checked-cast-test.cc

```diff
@@ -0,0 +1,58 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <type_traits>
+
+#include <gtest/gtest.h>
+
+#include "arrow/util/checked_cast.h"
+
+namespace arrow {
+
+#ifndef NDEBUG
```

Comment:
Is it possible to test release mode only? (presumably with the same classes, just different assertions)

Reply:
Let me look into it.

Reply:
The failure mode for release builds happens at compile time, so we can really only test the non-failure modes here.

Reply:
With the exception of the list array type bug, the bugs caught by `checked_cast` are all across the Python/C++ language boundary, FWIW.

Reply:
Yeah the bugs caught were two incorrect casts from python unions and nested list value builder appends for string and binary value types. They were working before only because of implicit assumptions about the methods being called and the data being modified.